SimGeneral/CaloAnalysis/plugins/MtdTruthAccumulator.cc

```diff
@@ -0,0 +1,937 @@
+  auto scHandle = event.put(std::move(output_.pSimClusters), "MergedMtdTruth");
+
+  // reserve for the best case scenario: already splitted
+  output_.pMtdSimLayerClusters->reserve(scHandle->size());
+  output_.pMtdSimTracksters->reserve(scHandle->size());
+
+  uint32_t SC_index = 0;
+  uint32_t LC_index = 0;
+  for (const auto &sc : *scHandle) {
+    auto const &hAndF = sc.hits_and_fractions();
+    auto const &hAndE = sc.hits_and_energies();
+    auto const &hAndT = sc.hits_and_times();
+    // create a vector with the indexes of the hits in the simCluster
+    std::vector<int> indexes(hAndF.size());
+    std::iota(indexes.begin(), indexes.end(), 0);
+    // sort the hits indexes based on subdetector, layer, module, then pixel row and column
+    std::sort(indexes.begin(), indexes.end(), [&](int a, int b) {
+      DetId aa = hAndF[a].first;
+      DetId bb = hAndF[b].first;
+      if (geomTools_.isETL(aa) != geomTools_.isETL(bb))
+        return (int)geomTools_.isETL(aa) < (int)geomTools_.isETL(bb);
+      else if (geomTools_.isETL(aa) and (geomTools_.getLayer(aa) != geomTools_.getLayer(bb)))
+        return geomTools_.getLayer(aa) < geomTools_.getLayer(bb);
+      else if (geomTools_.getModule(aa) != geomTools_.getModule(bb))
+        return geomTools_.getModule(aa) < geomTools_.getModule(aa);
+      else if (geomTools_.getPixelInModule(aa, m_detIdToPos[aa]).second ==
+                   (int)geomTools_.getPixelInModule(bb, m_detIdToPos[bb]).second and
+               std::abs(geomTools_.getPixelInModule(aa, m_detIdToPos[aa]).first -
+                        (int)geomTools_.getPixelInModule(bb, m_detIdToPos[bb]).first) < 2)
+        return geomTools_.getPixelInModule(aa, m_detIdToPos[aa]).first <
+               (int)geomTools_.getPixelInModule(bb, m_detIdToPos[bb]).first;
+      else
+        return geomTools_.getPixelInModule(aa, m_detIdToPos[aa]).second <
+               (int)geomTools_.getPixelInModule(bb, m_detIdToPos[bb]).second;
+    });
+
+    // now split the sc: loop on the sorted indexes and save the first hit in a
+    // temporary simCluster. If the following hit is in the same module and row (column),
+    // but next column (row) put it in the temporary simcluster as well, otherwise
+    // put the temporary simcluster in the collection and start creating a new one
+    std::vector<uint32_t> LC_indexes;
+    MtdSimLayerCluster tmpLC(sc.g4Tracks()[0]);
+    int prev = indexes[0];
+    DetId prevId(hAndF[prev].first);
+
+    float SimLCenergy = 0.;
+    float SimLCx = 0., SimLCy = 0., SimLCz = 0.;
+
+    auto push_back_hit = [&](const int &ind) {
+      tmpLC.addRecHitAndFraction(hAndF[ind].first, hAndF[ind].second);
+      tmpLC.addHitEnergy(hAndE[ind].second);
+      tmpLC.addHitTime(hAndT[ind].second);
+    };
+
+    auto update_clu_info = [&](const int &ind) {
+      SimLCenergy += convertUnitsTo(0.001_MeV, m_detIdToTotalSimEnergy[hAndF[ind].first]);
+      SimLCx +=
```

Comment:
@AuroraPerego this list of energy conversions from GeV to MeV should not be needed here. According to CMS conventions, see https://twiki.cern.ch/twiki/bin/view/CMS/CMSConventions#CMS_units , we should use GeV for energies, and MTD simHits are correctly stored with such a unit for energy. The issue is with MTD rechits, that historically have been erroneously recomputed in MeV. This is the reason of the energy conversions used in the MTD validation code. I would anyway prefer that we limit as much as possible the exception to the convention, which sooner or later I'd like to fix, and move such conversion only where strictly needed by direct comparison with reco hits/clusters (so far validation I would say). The fewer places where changes are done, the easier will be to fix it.

In case please clarify the absolute need for such a conversion in this place.

Reply:
I'll move back to GeV